[JENKINS-50164] Remove UI option to modify WS and builds locations

batmat opened the pull request:
Introduce system properties in lieu of the previous way using the UI for this.
Still leave the UI in readonly mode to ease checking configuration for admins.

Tests have been written to check using those system properties would work, even on an existing setup.

A *warning* log is now outputted when these system properties are used and changing the previous value, so that not only the help text warns about the fact there's no data migration.

See [JENKINS-50164](https://issues.jenkins-ci.org/browse/JENKINS-50164).

### TODO before/just after merge:

~~* [x] document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published. Requested by @daniel-beck  https://github.com/jenkins-infra/jenkins.io/pull/1510~~
* [x] Update the [sysprop standard page](https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties) with new syprops


<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* JENKINS-50164: Removed the option to modify _Build Record Root Directory_ and _Workspace Root Directory_ from the User Interface directly. Those are now only modifiable using system properties, refer to the [standard page for these values](https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties)  **TODO: write docs there once/if merged**. Remember that no data migration is handled when using those options on an existing setup. 

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees

batmat committed:
[JENKINS-50164] Remove UI option to modify WS and builds locations

Test (almost) only: checking the previous behaviour is preserved, and
the system property is accounted for when set.

batmat committed:
[JENKINS-50164] Remove UI option to modify WS and builds locations

Add necessary code to fix tests, and make the UI readonly.

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3040,25 @@ private void loadConfig() throws IOException {
+        if (newBuildsDir != null && !buildsDir.equals(newBuildsDir)) {
+            LOGGER.log(Level.WARNING, "Changing builds directories from {0} to {1}. Beware that no automated data migration will occur.",
+                       new String[]{buildsDir, newBuildsDir});
+            buildsDir = newBuildsDir;
```
And need to `save()`, I suspect.

batmat replied:
done

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5054,6 +5078,28 @@ private static void computeVersion(ServletContext context) {
+    private static final String OLD_DEFAULT_WORKSPACES_DIR = "${ITEM_ROOTDIR}/" + WORKSPACE_DIRNAME;
+
+    /**
+     * Default value for the workspace's directories layout.
```
`@see #workspaceDir`

batmat replied:
done, see below

jglick commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -5054,6 +5078,28 @@ private static void computeVersion(ServletContext context) {
+     * System property name to set {@link #buildsDir}.
+     * @see #getRawBuildsDir()
+     */
+    public static final String BUILDS_DIR_PROP = Jenkins.class.getName() + ".BUILDS_DIR";
```
I do not think these constants need to be `public`—we do not expect plugin Java code to refer to them.

batmat replied:
fixed

jglick commented on core/src/main/resources/jenkins/model/Jenkins/configure.jelly:
```diff
@@ -40,11 +40,11 @@ THE SOFTWARE.
       <f:advanced>
-        <f:entry field="rawWorkspaceDir" title="${%Workspace Root Directory}">
-          <f:textbox/>
+        <f:entry field="rawWorkspaceDir" title="${%Workspace Root Directory}" >
```
stray space

jglick commented on core/src/main/resources/jenkins/model/Jenkins/configure.jelly:
```diff
@@ -40,11 +40,11 @@ THE SOFTWARE.
         </f:entry>
         <f:entry field="rawBuildsDir" title="${%Build Record Root Directory}">
-          <f:textbox/>
+          <f:textbox readonly="readonly"/>
```
Setting the `readonly` attribute will block a web browser from editing this field, but `doConfigSubmit` is continuing to accept modifications from HTTP posts! I think you want to delete those two lines as well.

I would really just suggest deleting this `advanced` block, and print a warning to the system log if there is an override which is not matched by the system property.

batmat replied:
done it all

jglick commented on test/src/test/java/jenkins/model/JenkinsInitializationTest.java:
```diff
@@ -0,0 +1,98 @@
+    @Test
+    public void buildsDir() throws Exception {
+        loggerRule.record(Jenkins.class, Level.WARNING).capture(10);
+        story.addStep(new Statement() {
```
Use `then`, which takes a SAM that can be implemented via lambda.

batmat replied:
done

batmat committed:
Make package visible to avoid public usage

batmat committed:
workspaceDir & buildDir not modifiable anymore from the UI!

batmat committed:
Remove the whole UI for buildsDir and workspaceDir

batmat committed:
Lambda-ify

batmat committed:
Better javadoc

batmat committed:
Log always as INFO minimum, WARNING if values were changed

batmat committed:
save() if fields were touched

batmat commented:
@jglick I believe I addressed all your comments.

Only thing not done yet is checking how it works with https://issues.jenkins-ci.org/browse/JENKINS-21942 as requested in https://issues.jenkins-ci.org/browse/JENKINS-50164?focusedCommentId=332902&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-332902

batmat committed:
Move and refactor test as now the UI is not anymore

batmat commented:
Build succeeded, reviews are welcome. Thanks!

jtnord commented:
I am sceptical about this - esp because of a JIRA I am currently trying to track down - and also as this should be managed as CaC and not a System property.

Because as was said to me on a different PR No NOT another randomly hidden system property that alters behaviour / state.

daniel-beck commented:
@jtnord What alternative to having a UI option (which clearly leads to broken behavior) do you propose?

jtnord commented:
@daniel-beck 

> also as this should be managed as CaC and not a System property.

The UI option can lead to broken behaviour just as much as a System property. 
At least in the UI you can do `FormValidation` such that they can be warned before they submit...

daniel-beck commented:
@jtnord 

> The UI option can lead to broken behaviour just as much as a System property.

Not really, as it cannot easily be changed while Jenkins is running ([and I'm trying to prevent CasC doing something dumb in this comment](https://github.com/jenkinsci/configuration-as-code-plugin/issues/151#issuecomment-374012341)). Pulling the rug out from under builds in memory is pretty much unique to the UI option. (Also, no sane way around changing it via the UI, unless you really like XML).

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3040,36 @@ private void loadConfig() throws IOException {
+        }
+
+        if (mustSave) {
+            save();
```
seems scary as this is being called in a thread that is performing a load

batmat replied:
Recommended by @jglick in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r176423228

I'm not totally sure myself of the criticality of the risk, if there is one.

jtnord commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,241 @@
+ *
+ * @since TODO
+ */
+public class JenkinsBuildsAndWorkspacesDirectoriesTest {
```
Seems to be missing a test for the following:

* Customer build/workspace are set in existing Jenkins config (from version without this change)
* Jenkins is upgrades
* Jenkins config is saved
* configuration should not be reset - existing custom values should be retained until they set the system property

Because otherwise people are going to not do that step and mess things up.

And this implies there is no way to go back to default - once you have set the system property if you want to revert to the defaults you need to know what they are and set the system property to it (removing the system property should not change things)...

daniel-beck commented:
It is unclear why #3360 was closed instead of updated, but given @jtnord's comments it seems quite a few of the review comments there are still relevant here.

jtnord commented:
By broken I mean a user setting BUILD_ROOT=ITEM_ROOT.  [JENKINS-50335](https://issues.jenkins-ci.org/browse/JENKINS-50335)

jtnord commented:
A job can configure a custom workspace, not sure how that overlaps with this PR.

batmat commented:
> A job can configure a custom workspace, not sure how that overlaps with this PR.

IMO it does not. Or I mean, yes, this seems like an issue. But if so, was/is already existing so should be handled in a separate PR anyway.

batmat commented:
> It is unclear why #3360 was closed instead of updated

Yes. I hesitated, and went for a brand new PR as I didn't reuse anything from the previous one. The alternative was to force push, could have done that. (The 3rd alternative, to just add more commits and keep the outdated ones I really do not like, in such case this leads IMO lead to only messing things to use `git blame` afterwards).
 
> but given @jtnord's comments it seems quite a few of the review comments there are still relevant here.

Care to be more specific? I did spend a bit of time here to actually write migration tests, so apart from possibly addressing the specific use case noted by @jtnord, I'm not sure I see *exactly* what use case you think is not covered.

batmat committed:
Add to check removing sysprop keeps the previous value

Addressing @jtnord's concern/following use case:

* Custom build/workspace are set in existing Jenkins config (from version
  without this change)
* Jenkins is upgraded
* Jenkins config is saved
* configuration should not be reset - existing custom values should be
  retained until they set the system property

batmat commented:
@jtnord just pushed https://github.com/jenkinsci/jenkins/pull/3364/commits/774388598e1bf861330c29470484d7957d86fe69 that adds a test showing the use case you described in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r177042462 is tested. Could you please have another look?

@daniel-beck I've checked the other comments here and in #3360 and FWICT I've covered all of them by tests or answers.

Thanks a lot.

batmat committed:
Disable fromPreviousCustomSetup on Windows

Because I use restart and it's not supported there.

batmat committed:
Name sysprops more consistently with others

https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/configure.jelly:
```diff
@@ -39,14 +39,6 @@ THE SOFTWARE.
-        <f:entry field="rawBuildsDir" title="${%Build Record Root Directory}">
-          <f:textbox/>
-        </f:entry>
-      </f:advanced>
```
So we're removing the readonly UI after all?

batmat replied:
@daniel-beck I finally took the path Jesse advised in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r176424745

daniel-beck replied:
Preventing manipulated browser form submissions and not showing the current configuration are not necessarily the same thing.

This could easily be a table showing the current values and perhaps providing example paths.

jglick replied:
IMO: users should not be customizing this. If they do, they presumably have some reason and will be able to figure out example paths for themselves. If they are not customizing this, then they should have no reason to care. And the workspace root is only relevant for on-master builds, which you should never do to begin with.

daniel-beck replied:
@jglick 

> And the workspace root is only relevant for on-master builds, which you should never do to begin with.

Jenkins does it for you in matrix parents, pipelines, etc.

ci.j.io gets the path sanitization warning from `jenkins.branch.WorkspaceLocatorImpl locate` regularly due to the legacy setting, and there are no builds executing on master.

daniel-beck reviewed (dismissed):
This is dropping all the elaborate form validation I wrote to prevent terrible values, which seems much more likely with users accidentally not, or double, quoting the command args on the shell (i.e. `/builds` and `/workspace/` respectively for the 'default' values since the variables evaluate to the empty string).

What currently is [form validation](https://github.com/batmat/jenkins/blob/2c1fdd26990b417a4e57fdaaa7fb30ddc2a9be17/core/src/main/java/jenkins/model/Jenkins.java#L2140...L2178) should become errors preventing Jenkins startup.

Also, that unused code needs to then be removed.

batmat committed:
Check buildsDir value and fail Jenkins startup if wrong

batmat commented:
@daniel-beck I believe I addressed your comment in the last commit.

batmat committed:
Use released version of jenkins-test-harness

batmat commented:
@reviewbybees for more eyes.

@jtnord could you possibly please review again the addition?

@daniel-beck do you think you can have another look? 

Thanks

daniel-beck reviewed (approved):
In isolation, this change is now OK. I tested the behavior and what was form validation is now boot failures.

Since we're removing documentation from Jenkins, make sure to document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published.

Also, make sure the PR comment and changelog message are updated.

daniel-beck commented:
Would strongly prefer more reviews than just mine for this change.

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3001,86 @@ private void loadConfig() throws IOException {
+        if (replacedValue.contains("doCheckRawBuildsDir-Marker:foo")) {
+            // make sure platform can handle colon
+            try {
+                File tmp = File.createTempFile("Jenkins-doCheckRawBuildsDir", "foo:bar");
```
this would result in an illegal filename on windows as `:` is a reserved character - how are the tests passing??

batmat replied:
@jtnord no idea. Can check, but this test is a cut-n-paste from before, and looking around here, it does not seem it was in a test using `assume`. Checking on CI side.

jtnord replied:
I found it really hard to even grok what this whole method was trying to do. 
if it is cut and paste then _shrug_

daniel-beck replied:
We're trying to figure out whether there would be legal file names for a given item name (without having an item). This particular part implements the `ITEM_FULLNAME` vs. `ITEM_FULL_NAME` compatibility check.

batmat replied:
So, didn't dig into _why_ it works, but this test, from `JenkinsDescriptorTest` **is** currently successful in CI on master: https://ci.jenkins.io/job/Core/job/jenkins/job/master/867/testReport/jenkins.model/JenkinsDescriptorTest/Windows___Windows_Publishing___testBuildDirValidation/

![image](https://user-images.githubusercontent.com/223853/38996218-0ca4b09e-43eb-11e8-8bd3-16a64cf9d723.png)

daniel-beck replied:
@batmat There is no test for `ITEM_FULLNAME` due to the platform dependency, so this doesn't even happen during the test. This used to be basic form validation, so I didn't bother.

What's the problem here anyway? If the user wants `ITEM_FULLNAME`, that's only OK if the file system can handle colons. That's all this check does. Otherwise the error message will tell them to use `ITEM_FULL_NAME`, which escapes colons.

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3001,86 @@ private void loadConfig() throws IOException {
+    /*private*/ static void checkRawBuildsDir(String value) throws InvalidBuildsDir {
+
+        // do essentially what expandVariablesForDirectory does, without an Item
+        String replacedValue = expandVariablesForDirectory(value,
```
this method is utterly confusing

jtnord commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3001,86 @@ private void loadConfig() throws IOException {
+    }
+
+    /**
+     * @param value
```
useless javadoc.  what is it checking for - embedding bitcoins?

batmat replied:
Fixed

jtnord commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,364 @@
+    public void badValueForBuildsDir() {
+        story.then((rule) -> {
+            final List<String> badValues = Arrays.asList(
+                    "blah",
```
so there are no bad values for windows??  this seems to be unix specific without an assume and may fail but likely not for the expected reason.

jtnord commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,364 @@
+    @Test
+    public void fromPreviousCustomSetup() {
+
+        assumeFalse(Functions.isWindows()); // Default Windows lifecycle does not support restart.
```
NIT: it is much nicer to use `assumeFalse(" Default Windows lifecycle does not support restart.", Function.isWindows())` as then the reason for the skipped test will be reported.

batmat replied:
Fixed

jtnord reviewed (commented):
Seems ok apart from the `checkRawBuildsDir` method which is thoroughly confusing to me.

batmat committed:
Update javadoc & clarify parameter name

batmat committed:
Make the potential skip reason bubble up

batmat commented:
I fixed some of your comments @jtnord. Thanks for the review!

To /all I plan to merge this before the next release, i.e. on Sunday max if nobody objects in the meantime.

Thanks!

jglick commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,364 @@
+ * So this test class uses a {@link RestartableJenkinsRule} to check the behaviour of this sysprop being
+ * present or not between two restarts.
+ *
+ * @since TODO
```
No `@since` tag needed on a test.

jglick commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,364 @@
+        );
+
+        story.then(steps -> {
+            assertTrue(story.j.getInstance().isDefaultBuildDir());
```
The lambda takes a parameter. And `getInstance` is `static`. Conventionally

```java
story.then(r -> {
    assertTrue(r.jenkins.isDefaultBuildDir());
```

jglick reviewed (approved):
Not sure I understand all the implications, but removing a UI option is generally a good thing, and I am assuming you are testing this.

daniel-beck commented:
On-holding this until the following is addressed (or at least gets a reasonable response):

> Since we're removing documentation from Jenkins, make sure to document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published.
> 
> Also, make sure the PR comment and changelog message are updated.

batmat commented:
Working on the _on-holding_ reason now, hopefully to propose something very soon.

To loop back with a discussion on IRC with Daniel, I'm generally against squashing. So please just do not squash this one if you merge it. Thanks!

batmat commented:
Just filed the PR linked above to _on-hold_ this. Please review so that we can move forward here.

Thanks!

batmat commented:
Updated https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties?blah=eeek

batmat commented:
Going to merge this in the next 10 minutes if nobody objects. The PR was already approved, and I've fixed the reason for putting it On-Hold.

batmat commented:
For the record, the PR was built green previously. The current failure seems like an infra issue.

![image](https://user-images.githubusercontent.com/223853/39178782-db59cf80-47b2-11e8-84ee-45f34f9b456a.png)

![image](https://user-images.githubusercontent.com/223853/39178789-e160daae-47b2-11e8-9698-7bccf73d2c76.png)

KostyaSha commented:
Am i correctly understand:
Jenkins.java: `    private static final String DEFAULT_BUILDS_DIR = "${ITEM_ROOTDIR}/builds";` is a default value
Then `JenkinsRule` calls `Jenkins.loadTasks()` -> `Jenkins.loadConfig();`, by default `config.xml` is missing, so `checkRawBuildsDir(buildsDir)` gets `checkRawBuildsDir("${ITEM_ROOTDIR}/builds")` which expands `...../jobs/doCheckRawBuildsDir-Marker$foo/builds` and failed because it's not an absolute path?

KostyaSha commented:
@jglick @batmat @daniel-beck @oleg-nenashev ?

KostyaSha commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -3035,6 +3001,87 @@ private void loadConfig() throws IOException {
+
+    }
+
+    private void setBuildsAndWorkspacesDir() throws IOException, InvalidBuildsDir {
```
should it be covered with a test cases?

KostyaSha commented on test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java:
```diff
@@ -0,0 +1,364 @@
+                    "$JENKINS_HOME/foo/$ITEM_FULL_NAME",
+                    "${ITEM_ROOTDIR}/builds");
+
+            for (String goodValue : badValues) {
```
so is it good or bad?

jglick commented:
> when empty Jenkins is running does it hit into wrong default

[JENKINS-21942](https://issues.jenkins-ci.org/browse/JENKINS-21942)
